constraint/pkg/client/matchers.go

```diff
@@ -0,0 +1,158 @@
+package client
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/open-policy-agent/frameworks/constraint/pkg/client/errors"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/core/constraints"
+)
+
+// matcherKey uniquely identifies a Matcher.
+// For a given Constraint (uniquely identified by Kind/Name), there is at most
+// one Matcher for each Target.
+type matcherKey struct {
+	target string
+	kind   string
+	name   string
+}
+
+// constraintMatchers tracks the Matchers for each Constraint.
+// Filters Constraints relevant to a passed review.
+type constraintMatchers struct {
+	// matchers is the set of Constraint matchers by their Target, Kind, and Name.
+	// matchers is a map from Target to a map from Kind to a map from Name to matcher.
+	matchers map[string]map[string]map[string]constraints.Matcher
+
+	mtx sync.RWMutex
+}
+
+// Add inserts the Matcher for the Constraint with kind and name.
+// Replaces the current Matcher if one already exists.
+func (c *constraintMatchers) Add(key matcherKey, matcher constraints.Matcher) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if c.matchers == nil {
+		c.matchers = make(map[string]map[string]map[string]constraints.Matcher)
+	}
+
+	targetMatchers := c.matchers[key.target]
+	if targetMatchers == nil {
+		targetMatchers = make(map[string]map[string]constraints.Matcher)
+	}
+
+	kindMatchers := targetMatchers[key.kind]
+	if kindMatchers == nil {
+		kindMatchers = make(map[string]constraints.Matcher)
+	}
+
+	kindMatchers[key.name] = matcher
+	targetMatchers[key.kind] = kindMatchers
+	c.matchers[key.target] = targetMatchers
+}
+
+// Remove deletes the Matcher for the Constraint with kind and name.
+// Returns normally if no entry for the Constraint existed.
+func (c *constraintMatchers) Remove(key matcherKey) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if len(c.matchers) == 0 {
+		return
+	}
+
+	targetMatchers := c.matchers[key.target]
+	if len(targetMatchers) == 0 {
+		return
+	}
+
+	kindMatchers := targetMatchers[key.kind]
+	if len(kindMatchers) == 0 {
+		return
+	}
+
+	delete(kindMatchers, key.name)
+
+	// Remove empty parents to avoid memory leaks.
+	if len(kindMatchers) == 0 {
+		delete(targetMatchers, key.kind)
+	} else {
+		targetMatchers[key.kind] = kindMatchers
+	}
+
+	if len(targetMatchers) == 0 {
+		delete(c.matchers, key.target)
+	} else {
+		c.matchers[key.target] = targetMatchers
+	}
+}
+
+// RemoveAll removes all Matchers for Constraints with kind.
+// Returns normally if no entry for the kind existed.
+func (c *constraintMatchers) RemoveAll(kind string) {
```

Comment:
nit: `RemoveKind`? To disambiguate vs. purging all matchers?

Reply:
Done.